From e41b2b10e637920665cf590bc170407d414ecf32 Mon Sep 17 00:00:00 2001
From: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
Date: Wed, 3 Nov 2021 18:34:11 +0300
Subject: [PATCH 041/249] zdtm.py: make tests with --link_remap exclusive

We see that tests mntns_ghost01 and unlink_fstat03 can run
simultaneousely and thus the former sees leftover link_remap.* files in
the test directory created by the latter, and the latter is still
running so it's ok to have link_remap.* at this point.

Let's implicitly make all --link-remap tests exclusive (not running in
parallel).

Fixes: #1633

Signed-off-by: Pavel Tikhomirov <ptikhomirov@virtuozzo.com>
---
 test/zdtm.py | 19 +++++++++++++++++--
 1 file changed, 17 insertions(+), 2 deletions(-)

diff --git a/test/zdtm.py b/test/zdtm.py
index fc7b8a183..b62136e96 100755
--- a/test/zdtm.py
+++ b/test/zdtm.py
@@ -1997,7 +1997,22 @@ class Launcher:
             raise Exception("The kernel is tainted: %r (%r)" %
                             (taint, self.__taint))
 
-        if test_flag(desc, 'excl'):
+        '''
+        The option --link-remap allows criu to hardlink open files back to the
+        file-system on dump (should be removed on restore) and we have a sanity
+        check in check_visible_state that they were actually removed at least
+        from the root test directory after restore.
+
+        As zdtm runs all tests from the same cwd (e.g.: test/zdtm/static) in
+        parallel, hardlinks from one test can mess up with sanity checks of
+        another test or even one test can by mistake use hardlinks created by
+        another test which is even worse.
+
+        So let's make all tests using --link-remap option non parallel.
+        '''
+        link_remap_excl = '--link-remap' in desc.get('opts', '').split() + desc.get('dopts', '').split() + desc.get('ropts', '').split()
+
+        if test_flag(desc, 'excl') or link_remap_excl:
             self.wait_all()
 
         self.__nr += 1
@@ -2030,7 +2045,7 @@ class Launcher:
             "start": time.time()
         }
 
-        if test_flag(desc, 'excl'):
+        if test_flag(desc, 'excl') or link_remap_excl:
             self.wait()
 
     def __wait_one(self, flags):
-- 
2.35.1

